main/src/ca/uwaterloo/flix/language/phase/Lowering.scala

```diff
@@ -1688,6 +1711,229 @@ object Lowering {
+    val predSymVar = Symbol.freshVarSym("predSym", BoundBy.FormalParam, loc)
+    val termsVar = Symbol.freshVarSym("terms", BoundBy.FormalParam, loc)
+    mkLambdaExp(predSymVar, Types.PredSym,
+      mkLambdaExp(termsVar, Types.VectorOfBoxed,
+        mkExtVarOuterBody(preds, predSymVar, termsVar, tpe, loc),
+        tpe, Type.Pure, loc
+      ),
+      Type.mkPureArrow(Types.VectorOfBoxed, tpe, loc), Type.Pure, loc
+    )
+  }
+
+  /**
+    * Returns the `TypedAst` lambda expression
+    * {{{
+    *   paramName -> exp
+    * }}}
+    * where `"paramName" == param.text` and `exp` has type `expType` and effect `eff`.
+    */
+  private def mkLambdaExp(param: Symbol.VarSym, paramTpe: Type, exp: TypedAst.Expr, expTpe: Type, eff: Type, loc: SourceLocation): TypedAst.Expr =
+    TypedAst.Expr.Lambda(
+      TypedAst.FormalParam(TypedAst.Binder(param, paramTpe), Modifiers.Empty, paramTpe, TypeSource.Ascribed, loc),
+      exp,
+      Type.mkArrowWithEffect(paramTpe, eff, expTpe, loc),
+      loc
+    )
+
+  /**
+    Returns the `TypedAst` match expression
+    * {{{
+    *   match predSym {
+    *     case PredSym.PredSym(name, _) => match name {
+    *       case "P1" => xvar P1(unbox(Vector.get(0, terms)), unbox(Vector.get(1, terms)), ...)
+    *       case "P2" => xvar P2(unbox(Vector.get(0, terms)), unbox(Vector.get(1, terms)), ...)
+    *       ...
+    *       case _ => bug!("name is not a predicate")
+    *     }
+    *   }
+    * }}}
+    * where `P1, P2, ...` are in `preds` with their respective term types, `"predSym" == predSymVar.text`
+    * and `"terms" == termsVar.text`.
+    */
+  private def mkExtVarOuterBody(preds: List[(Name.Pred, List[Type])], predSymVar: Symbol.VarSym, termsVar: Symbol.VarSym, tpe: Type, loc: SourceLocation)(implicit scope: Scope, flix: Flix): TypedAst.Expr = {
+    val name = Symbol.freshVarSym(Name.Ident("name", loc), BoundBy.Pattern)
+    TypedAst.Expr.Match(
+      exp = TypedAst.Expr.Var(predSymVar, Types.PredSym, loc),
+      rules = List(
+        TypedAst.MatchRule(
+          pat = TypedAst.Pattern.Tag(
+            CaseSymUse(Symbol.mkCaseSym(Enums.PredSym, Name.Ident("PredSym", loc)), loc),
+            List(
+              TypedAst.Pattern.Var(TypedAst.Binder(name, Type.Str), Type.Str, loc),
+              TypedAst.Pattern.Wild(Type.Int64, loc)
+            ),
+            Types.PredSym, loc
+          ),
+          guard = None, exp = mkExtVarInnerBody(preds, name, termsVar, tpe, loc), loc
+        )
+      ),
+      tpe = tpe, eff = Type.Pure, loc
+    )
+  }
+
+
+  /**
+    * Returns the `TypedAst` match expression
+    * {{{
+    *   match name {
+    *     case "P1" => xvar P1(unbox(Vector.get(0, terms)), unbox(Vector.get(1, terms)), ...)
+    *     case "P2" => xvar P2(unbox(Vector.get(0, terms)), unbox(Vector.get(1, terms)), ...)
+    *     ...
+    *     case _ => bug!("name is not a predicate")
```

Comment:
This case should just be omitted. The Simplifier will insert a MatchError rule itself.

This also avoids depending on bug! from the standard library.